File: src/worker/normalize/sb_urls.go
// Copyright 2016 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This file contains code copied from github.com/google/safebrowsing
// (urls.go, Apache-2.0). MODIFICATIONS relative to the original:
//
//  1. Google's custom parseURL/parseHost/canonicalURL are NOT copied. Parsing is
//     done by the host language's URL parser instead (net/url here,
//     WHATWG `new URL()` in the Worker) so the Go and TypeScript sides agree.
//     See canonical.go.
//  2. escapePath escapes '?' in addition to Google's set. Google unescapes the
//     whole URL before splitting, so a decoded '?' always starts the query;
//     we split first, so a '?' decoded inside the path must be re-escaped or
//     the canonical string would be ambiguous. escapeQuery is Google's set.
//  3. generateLookupHosts/generateLookupPaths take already-parsed parts instead
//     of a URL string, and return no error.
//  4. Unused helpers (split, getScheme, canonicalPath, ValidURL, generateHashes)
//     are omitted.
//
// Any change here must be mirrored in worker/src/sb_urls.ts.
package normalize

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// isHex reports whether c is a hexadecimal character.
func isHex(c byte) bool {
	switch {
	case '0' <= c && c <= '9':
		return true
	case 'a' <= c && c <= 'f':
		return true
	case 'A' <= c && c <= 'F':
		return true
	}
	return false
}

// unhex converts a hexadecimal character to byte value in 0..15, inclusive.
func unhex(c byte) byte {
	switch {
	case '0' <= c && c <= '9':
		return c - '0'
	case 'a' <= c && c <= 'f':
		return c - 'a' + 10
	case 'A' <= c && c <= 'F':
		return c - 'A' + 10
	}
	return 0
}

// isUnicode reports whether s is a Unicode string.
func isUnicode(s string) bool {
	for _, c := range []byte(s) {
		// For legacy reasons, 0x80 is not considered a Unicode character.
		if c > 0x80 {
			return true
		}
	}
	return false
}

// escape returns the percent-encoded form of the string s.
// MODIFIED: takes the extra character set as a parameter (see escapePath).
func escape(s string, extra string) string {
	var b strings.Builder
	b.Grow(len(s))
	for _, c := range []byte(s) {
		if c < 0x20 || c >= 0x7f || c == ' ' || c == '#' || c == '%' ||
			strings.IndexByte(extra, c) >= 0 {
			fmt.Fprintf(&b, "%%%02x", c)
		} else {
			b.WriteByte(c)
		}
	}
	return b.String()
}

func escapePath(s string) string  { return escape(s, "?") }
func escapeQuery(s string) string { return escape(s, "") }

// unescape returns the decoded form of a percent-encoded string s.
func unescape(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for len(s) > 0 {
		if len(s) >= 3 && s[0] == '%' && isHex(s[1]) && isHex(s[2]) {
			b.WriteByte(unhex(s[1])<<4 | unhex(s[2]))
			s = s[3:]
		} else {
			b.WriteByte(s[0])
			s = s[1:]
		}
	}
	return b.String()
}

// recursiveUnescape unescapes the string s recursively until it cannot be
// unescaped anymore. It reports an error if the unescaping process seemed to
// have no end.
func recursiveUnescape(s string) (string, error) {
	const maxDepth = 1024
	for i := 0; i < maxDepth; i++ {
		t := unescape(s)
		if t == s {
			return s, nil
		}
		s = t
	}
	return "", errors.New("normalize: unescaping is too recursive")
}

// parseIPAddress returns the canonical dotted-decimal form of iphostname, or ""
// if it is not an IP address. It handles the hex, octal and short forms that
// legacy resolvers accept (0x7f000001, 127.1, "10.192.95.89 xy").
func parseIPAddress(iphostname string) string {
	// The Windows resolver allows a 4-part dotted decimal IP address to have a
	// space followed by any old rubbish, so long as the total length of the
	// string doesn't get above 15 characters. So, "10.192.95.89 xy" is
	// resolved to 10.192.95.89. If the string length is greater than 15
	// characters, e.g. "10.192.95.89 xy.wildcard.example.com", it will be
	// resolved through DNS.
	if len(iphostname) <= 15 {
		match := trailingSpaceRegexp.FindString(iphostname)
		if match != "" {
			iphostname = strings.TrimSpace(match)
		}
	}
	if !possibleIPRegexp.MatchString(iphostname) {
		return ""
	}

	parts := strings.Split(iphostname, ".")
	if len(parts) > 4 {
		return ""
	}
	ss := make([]string, len(parts))
	for i, n := range parts {
		if i == len(parts)-1 {
			ss[i] = canonicalNum(n, 5-len(parts))
		} else {
			ss[i] = canonicalNum(n, 1)
		}
		if ss[i] == "" {
			return ""
		}
	}
	return strings.Join(ss, ".")
}

// canonicalNum parses s as an integer and attempts to encode it as a '.'
// separated string where each element is the base-10 encoded value of each byte
// for the corresponding number, starting with the MSB. The result is one that
// is usable as an IP address.
//
// For example:
//
//	s:"01234",      n:2 => "2.156"
//	s:"0x10203040", n:4 => "16.32.48.64"
func canonicalNum(s string, n int) string {
	if n <= 0 || n > 4 {
		return ""
	}
	v, err := strconv.ParseUint(s, 0, 32)
	if err != nil {
		return ""
	}
	ss := make([]string, n)
	for i := n - 1; i >= 0; i-- {
		ss[i] = strconv.Itoa(int(v) & 0xff)
		v = v >> 8
	}
	return strings.Join(ss, ".")
}

// generateLookupHosts returns a list of host-suffixes for the parsed host.
//
// Safe Browsing policy asks to generate lookup hosts for the URL. Those are
// formed by the domain and also up to 4 hostname suffixes. The last component
// or sometimes the pair isn't examined alone, since it's the TLD or country
// code.
//
// Note that we do not need to be clever about stopping at the "real" TLD.
// We just check a few extra components regardless.
func generateLookupHosts(host string, isIP bool) []string {
	const maxHostComponents = 7

	if isIP {
		return []string{host}
	}

	hostComponents := strings.Split(host, ".")
	numComponents := len(hostComponents) - maxHostComponents
	if numComponents < 1 {
		numComponents = 1
	}

	hosts := []string{host}
	for i := numComponents; i < len(hostComponents)-1; i++ {
		hosts = append(hosts, strings.Join(hostComponents[i:], "."))
	}
	return hosts
}

// generateLookupPaths returns a list of path-prefixes for the parsed path/query.
func generateLookupPaths(path, query string) []string {
	const maxPathComponents = 4

	paths := []string{"/"}
	var pathComponents []string
	for _, p := range strings.Split(path, "/") {
		if p != "" {
			pathComponents = append(pathComponents, p)
		}
	}

	numComponents := len(pathComponents)
	if numComponents > maxPathComponents {
		numComponents = maxPathComponents
	}
	for i := 1; i < numComponents; i++ {
		paths = append(paths, "/"+strings.Join(pathComponents[:i], "/")+"/")
	}
	if path != "/" {
		paths = append(paths, path)
	}
	if len(query) > 0 {
		paths = append(paths, path+"?"+query)
	}
	return paths
}

File: client/src/sb_urls.ts
// Copyright 2016 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// TypeScript port of the helpers in github.com/google/safebrowsing/urls.go
// (Apache-2.0). Line-for-line mirror of internal/normalize/sb_urls.go — see the
// MODIFICATIONS note there. Change both files together or lookups stop matching.

/** Percent-encode. `extra` holds characters escaped beyond Google's set. */
export function escape(s: string, extra: string): string {
  let out = '';
  for (const c of bytesOf(s)) {
    if (c < 0x20 || c >= 0x7f || c === 0x20 || c === 0x23 || c === 0x25 || extra.includes(String.fromCharCode(c))) {
      out += '%' + c.toString(16).padStart(2, '0'); // lowercase, like Go's %02x
    } else {
      out += String.fromCharCode(c);
    }
  }
  return out;
}

export const escapePath = (s: string) => escape(s, '?');
export const escapeQuery = (s: string) => escape(s, '');

/** Decode one percent-encoding pass. Operates on bytes, like the Go version. */
export function unescape(s: string): string {
  let out = '';
  for (let i = 0; i < s.length; i++) {
    if (i + 2 < s.length && s[i] === '%' && isHex(s[i + 1]) && isHex(s[i + 2])) {
      out += String.fromCharCode((unhex(s[i + 1]) << 4) | unhex(s[i + 2]));
      i += 2;
    } else {
      out += s[i];
    }
  }
  return out;
}

/** Decode until stable, so %2520 and friends cannot hide a separator. */
export function recursiveUnescape(s: string): string {
  const maxDepth = 1024;
  for (let i = 0; i < maxDepth; i++) {
    const t = unescape(s);
    if (t === s) return s;
    s = t;
  }
  throw new Error('normalize: unescaping is too recursive');
}

const possibleIPRe = /^(?:0x[0-9a-f]+|[0-9.])+$/i;
const trailingSpaceRe = /^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}) /;

/**
 * Canonical dotted-decimal form of an IP host, or '' if it is not one.
 * Handles the hex, octal and short forms legacy resolvers accept
 * (0x7f000001, 127.1, "10.192.95.89 xy").
 */
export function parseIPAddress(iphostname: string): string {
  if (iphostname.length <= 15) {
    const m = trailingSpaceRe.exec(iphostname);
    if (m) iphostname = m[0].trim();
  }
  if (!possibleIPRe.test(iphostname)) return '';

  const parts = iphostname.split('.');
  if (parts.length > 4) return '';

  const ss: string[] = [];
  for (let i = 0; i < parts.length; i++) {
    const n = i === parts.length - 1 ? 5 - parts.length : 1;
    const s = canonicalNum(parts[i], n);
    if (s === '') return '';
    ss.push(s);
  }
  return ss.join('.');
}

/** "0x10203040", 4 => "16.32.48.64";  "01234", 2 => "2.156" */
export function canonicalNum(s: string, n: number): string {
  if (n <= 0 || n > 4) return '';
  const v = parseUintBase0(s);
  if (v === null) return '';

  const ss = new Array<string>(n);
  let rest = v;
  for (let i = n - 1; i >= 0; i--) {
    ss[i] = String(rest % 256);
    rest = Math.floor(rest / 256);
  }
  return ss.join('.');
}

/** Go's strconv.ParseUint(s, 0, 32): 0x hex, leading-0 octal, else decimal. */
function parseUintBase0(s: string): number | null {
  let v: number;
  if (/^0[xX][0-9a-fA-F]+$/.test(s)) v = parseInt(s.slice(2), 16);
  else if (/^0[0-7]*$/.test(s)) v = parseInt(s, 8);
  else if (/^[1-9][0-9]*$/.test(s)) v = parseInt(s, 10);
  else return null;
  return v > 0xffffffff ? null : v;
}

/** Host plus up to 4 parent suffixes (never the bare TLD). */
export function generateLookupHosts(host: string, isIP: boolean): string[] {
  const maxHostComponents = 7;
  if (isIP) return [host];

  const c = host.split('.');
  const start = Math.max(1, c.length - maxHostComponents);

  const hosts = [host];
  for (let i = start; i < c.length - 1; i++) hosts.push(c.slice(i).join('.'));
  return hosts;
}

/** '/', then up to 3 parent directories, then the full path and path+query. */
export function generateLookupPaths(path: string, query: string): string[] {
  const maxPathComponents = 4;

  const paths = ['/'];
  const comps = path.split('/').filter((p) => p !== '');
  const num = Math.min(comps.length, maxPathComponents);
  for (let i = 1; i < num; i++) paths.push('/' + comps.slice(0, i).join('/') + '/');
  if (path !== '/') paths.push(path);
  if (query.length > 0) paths.push(path + '?' + query);
  return paths;
}

// --- helpers ---

const utf8 = new TextEncoder();

/** Byte values of s. Chars above 0xFF are UTF-8 encoded, as Go strings are. */
function bytesOf(s: string): number[] {
  const out: number[] = [];
  for (const ch of s) {
    const c = ch.codePointAt(0)!;
    if (c <= 0xff) out.push(c);
    else for (const b of utf8.encode(ch)) out.push(b);
  }
  return out;
}

function isHex(c: string): boolean {
  return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
}

function unhex(c: string): number {
  if (c >= '0' && c <= '9') return c.charCodeAt(0) - 48;
  if (c >= 'a' && c <= 'f') return c.charCodeAt(0) - 87;
  if (c >= 'A' && c <= 'F') return c.charCodeAt(0) - 55;
  return 0;
}
